src/File.php

```
@@ -0,0 +1,397 @@
+        }
+        
+        // 'aes-128-ctr' unless someone mucked with the config in Core.php
+        $method = self::CIPHER.'-'.self::CIPHER_MODE;
+        
+        /**
+         *  Let's split our keys
+         */
+        
+        // $ekey -- Encryption Key -- used for AES
+        $ekey = self::HKDF(
+            self::HASH_FUNCTION,
+            $key,
+            self::KEY_BYTE_SIZE,
+            self::ENCRYPTION_INFO
+        );
+        
+        // $akey -- Authentication Key -- used for HMAC
+        $akey = self::HKDF(
+            self::HASH_FUNCTION,
+            $key,
+            self::KEY_BYTE_SIZE,
+            self::AUTHENTICATION_INFO
+        );
+        
+        /**
+         *  Generate a random initialization vector.
+         */
+        self::ensureFunctionExists("openssl_cipher_iv_length");
+        $ivsize = openssl_cipher_iv_length($method);
+        if ($ivsize === FALSE || $ivsize <= 0) {
+            throw new Ex\CannotPerformOperation('Improper IV size');
+        }
+        $iv = self::secureRandom($ivsize);
+        
+        /**
+         * First let's write our IV to the first N blocks of the output file
+         */
+        \fwrite($outputHandle, $iv, $ivsize);
+        
+        /**
+         * We're going to initialize a HMAC-SHA256 with the given $akey
+         * and update it with each ciphertext chunk
+         */
+        $hmac = \hash_init(self::HASH_FUNCTION, HASH_HMAC, $akey);
+        // We operate on $thisIv using a hash-based PRF derived from the initial
+        // IV for the first block
+        $thisIv = $iv;
+        
+        // How much do we increase the counter after each buffered encryption to prevent nonce reuse
+        $inc = self::BUFFER / self::BLOCK_SIZE;
+        
+        /**
+         * Let's MAC our IV/nonce
+         */
+        \hash_update($hmac, $iv);
+        
+        /**
+         * Iterate until we reach the end of the input file
+         */
+        while (!\feof($inputHandle)) {
```

I/O can fail at any time, for example if the disk got yanked out halfway through processing the file, or possibly if the adversary controlled the file permissions and changed them at just the right time. If it's an error rather than actually reaching the end of the file, we should raise an exception.